File: apps/app/src/lib/overview-summary.ts
/**
 * Pure composition logic for the overview summary card: portfolio vs market
 * 24h change, aggregated news sentiment, and the human-readable sentence
 * segments rendered above the activity feed.
 *
 * Segments are structured (`text` | `pct`) so the card can inline percent
 * badges without any string parsing.
 */

import type { BreadthStats } from "./overview-daily-brief";

/**
 * Value-weighted 24h change: Σ(value·pct) / Σ(value) over coins that have
 * both a positive USD value and a finite change percent. Coins missing a
 * quote are excluded from the denominator too, so partial quote maps don't
 * drag the average toward zero.
 */
export function computeValueWeighted24hChangePct(
  rows: ReadonlyArray<{ valueUsd: number; changePct: number | null }>,
): number | null {
  let weightedSum = 0;
  let totalValueUsd = 0;
  for (const row of rows) {
    if (!Number.isFinite(row.valueUsd) || row.valueUsd <= 0) continue;
    if (row.changePct === null || !Number.isFinite(row.changePct)) continue;
    weightedSum += row.valueUsd * row.changePct;
    totalValueUsd += row.valueUsd;
  }
  if (totalValueUsd <= 0) return null;
  return weightedSum / totalValueUsd;
}

/** First→last percent change of a series; null when it can't be computed. */
export function computeSeriesChangePct(
  points: ReadonlyArray<{ time: number; value: number }>,
): number | null {
  if (points.length < 2) return null;
  const startValue = points[0]?.value;
  const endValue = points[points.length - 1]?.value;
  if (
    startValue === undefined ||
    !Number.isFinite(startValue) ||
    startValue <= 0
  )
    return null;
  if (endValue === undefined || !Number.isFinite(endValue)) return null;
  return ((endValue - startValue) / startValue) * 100;
}

export type NewsSentimentLean = "bullish" | "bearish" | "mixed" | "quiet";

export interface NewsSentimentSummary {
  bullish: number;
  bearish: number;
  neutral: number;
  total: number;
  lean: NewsSentimentLean;
}

/**
 * Tally scored articles into a lean. A lean needs a margin (strictly more
 * than one extra story) so a 2-vs-1 tape reads as "mixed", not "bullish".
 */
export function aggregateNewsSentiment(
  rows: ReadonlyArray<{ sentiment: "bullish" | "bearish" | "neutral" | null }>,
): NewsSentimentSummary {
  let bullish = 0;
  let bearish = 0;
  let neutral = 0;
  for (const row of rows) {
    if (row.sentiment === "bullish") bullish++;
    else if (row.sentiment === "bearish") bearish++;
    else if (row.sentiment === "neutral") neutral++;
  }
  const total = bullish + bearish + neutral;
  const lean: NewsSentimentLean =
    total === 0
      ? "quiet"
      : bullish > bearish + 1
        ? "bullish"
        : bearish > bullish + 1
          ? "bearish"
          : "mixed";
  return { bullish, bearish, neutral, total, lean };
}

export type SummarySegment =
  | { kind: "text"; text: string }
  | { kind: "pct"; value: number };

/** Portfolio and market are "in line" within this many percentage points. */
const IN_LINE_BAND_PP = 0.25;

function directionWord(pct: number): "up" | "down" | "flat" {
  if (pct > 0) return "up";
  if (pct < 0) return "down";
  return "flat";
}

function text(value: string): SummarySegment {
  return { kind: "text", text: value };
}

function pct(value: number): SummarySegment {
  return { kind: "pct", value };
}

function buildPerformanceSentence(args: {
  portfolioChangePct: number | null;
  marketChangePct: number | null;
}): SummarySegment[] {
  const { portfolioChangePct, marketChangePct } = args;

  if (portfolioChangePct !== null && marketChangePct !== null) {
    const diff = portfolioChangePct - marketChangePct;
    const relation =
      Math.abs(diff) < IN_LINE_BAND_PP
        ? "in line with"
        : diff > 0
          ? "ahead of"
          : "behind";
    return [
      text(`Your portfolio is ${directionWord(portfolioChangePct)} `),
      pct(portfolioChangePct),
      text(` today, ${relation} the market's `),
      pct(marketChangePct),
      text("."),
    ];
  }

  if (portfolioChangePct !== null) {
    return [
      text(`Your portfolio is ${directionWord(portfolioChangePct)} `),
      pct(portfolioChangePct),
      text(" over the last 24h; the market benchmark is still warming up."),
    ];
  }

  if (marketChangePct !== null) {
    return [
      text(`The broader market is ${directionWord(marketChangePct)} `),
      pct(marketChangePct),
      text(" over the last 24h."),
    ];
  }

  return [];
}

function buildToneSentence(args: {
  sentiment: NewsSentimentSummary | null;
  breadth: BreadthStats | null;
}): SummarySegment[] {
  const { sentiment, breadth } = args;

  if (sentiment && sentiment.total > 0 && sentiment.lean !== "quiet") {
    if (sentiment.lean === "bullish") {
      return [
        text(
          `News flow around your coins leans bullish (${sentiment.bullish} of ${sentiment.total} ${sentiment.total === 1 ? "story" : "stories"}).`,
        ),
      ];
    }
    if (sentiment.lean === "bearish") {
      return [
        text(
          `News flow around your coins leans bearish (${sentiment.bearish} of ${sentiment.total} ${sentiment.total === 1 ? "story" : "stories"}).`,
        ),
      ];
    }
    return [
      text(
        `News flow around your coins is mixed across ${sentiment.total} ${sentiment.total === 1 ? "story" : "stories"}.`,
      ),
    ];
  }

  if (breadth) {
    const total = breadth.advancers + breadth.decliners + breadth.flat;
    if (total > 0) {
      if (breadth.advancers > breadth.decliners) {
        return [
          text(
            `${breadth.advancers} of ${total} ${total === 1 ? "coin is" : "coins are"} trading higher over the last 24h.`,
          ),
        ];
      }
      if (breadth.decliners > breadth.advancers) {
        return [
          text(
            `${breadth.decliners} of ${total} ${total === 1 ? "coin is" : "coins are"} trading lower over the last 24h.`,
          ),
        ];
      }
      return [
        text(
          "Your coins are evenly split between gainers and losers over the last 24h.",
        ),
      ];
    }
  }

  return [];
}

/** Merge adjacent text segments so renderers get the fewest nodes possible. */
function mergeAdjacentText(segments: SummarySegment[]): SummarySegment[] {
  const merged: SummarySegment[] = [];
  for (const segment of segments) {
    const previous = merged[merged.length - 1];
    if (segment.kind === "text" && previous?.kind === "text") {
      merged[merged.length - 1] = text(previous.text + segment.text);
      continue;
    }
    merged.push(segment);
  }
  return merged;
}

/**
 * Compose the summary prose: a performance sentence (portfolio vs market,
 * degrading to whichever side is available) followed by a tone sentence
 * (news sentiment, falling back to breadth). Empty when nothing is known.
 */
export function buildOverviewSummarySegments(args: {
  portfolioChangePct: number | null;
  marketChangePct: number | null;
  sentiment: NewsSentimentSummary | null;
  breadth: BreadthStats | null;
}): SummarySegment[] {
  const performance = buildPerformanceSentence(args);
  const tone = buildToneSentence(args);

  const segments: SummarySegment[] = [...performance];
  if (performance.length > 0 && tone.length > 0) segments.push(text(" "));
  segments.push(...tone);
  return mergeAdjacentText(segments);
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Overview/OverviewSummary.swift
import Foundation

/// Port of `lib/overview-summary.ts`.
public enum OverviewSummary {
  public struct ValueRow: Sendable { public var valueUsd: Double; public var changePct: Double?; public init(valueUsd: Double, changePct: Double?) { self.valueUsd = valueUsd; self.changePct = changePct } }

  /// Σ(value·pct) / Σ(value) over coins with a positive value and a finite change.
  public static func valueWeighted24hChangePct(_ rows: [ValueRow]) -> Double? {
    var weighted = 0.0, total = 0.0
    for r in rows {
      guard r.valueUsd.isFinite, r.valueUsd > 0, let pct = r.changePct, pct.isFinite else { continue }
      weighted += r.valueUsd * pct
      total += r.valueUsd
    }
    return total > 0 ? weighted / total : nil
  }

  public static func seriesChangePct(_ points: [TimePoint]) -> Double? {
    guard points.count >= 2 else { return nil }
    let start = points[0].value, end = points[points.count - 1].value
    guard start.isFinite, start > 0, end.isFinite else { return nil }
    return (end - start) / start * 100
  }

  public enum Sentiment: String, Sendable, Codable { case bullish, bearish, neutral }
  public enum Lean: String, Sendable { case bullish, bearish, mixed, quiet }

  public struct SentimentSummary: Sendable, Hashable {
    public var bullish: Int, bearish: Int, neutral: Int, total: Int
    public var lean: Lean
  }

  public static func aggregateNewsSentiment(_ rows: [Sentiment?]) -> SentimentSummary {
    var b = 0, be = 0, n = 0
    for r in rows { switch r { case .bullish?: b += 1; case .bearish?: be += 1; case .neutral?: n += 1; case nil: break } }
    let total = b + be + n
    let lean: Lean = total == 0 ? .quiet : (b > be + 1 ? .bullish : (be > b + 1 ? .bearish : .mixed))
    return SentimentSummary(bullish: b, bearish: be, neutral: n, total: total, lean: lean)
  }

  public enum Segment: Sendable, Hashable {
    case text(String)
    case pct(Double)
  }

  static let inLineBandPP = 0.25

  static func direction(_ pct: Double) -> String { pct > 0 ? "up" : (pct < 0 ? "down" : "flat") }

  static func performance(portfolio: Double?, market: Double?) -> [Segment] {
    if let p = portfolio, let m = market {
      let diff = p - m
      let relation = abs(diff) < inLineBandPP ? "in line with" : (diff > 0 ? "ahead of" : "behind")
      return [.text("Your portfolio is \(direction(p)) "), .pct(p), .text(" today, \(relation) the market's "), .pct(m), .text(".")]
    }
    if let p = portfolio {
      return [.text("Your portfolio is \(direction(p)) "), .pct(p), .text(" over the last 24h; the market benchmark is still warming up.")]
    }
    if let m = market {
      return [.text("The broader market is \(direction(m)) "), .pct(m), .text(" over the last 24h.")]
    }
    return []
  }

  static func tone(sentiment: SentimentSummary?, breadth: BreadthStats?) -> [Segment] {
    if let s = sentiment, s.total > 0, s.lean != .quiet {
      let stories = s.total == 1 ? "story" : "stories"
      switch s.lean {
      case .bullish: return [.text("News flow around your coins leans bullish (\(s.bullish) of \(s.total) \(stories)).")]
      case .bearish: return [.text("News flow around your coins leans bearish (\(s.bearish) of \(s.total) \(stories)).")]
      default: return [.text("News flow around your coins is mixed across \(s.total) \(stories).")]
      }
    }
    if let b = breadth, b.total > 0 {
      let total = b.total
      let coins = total == 1 ? "coin is" : "coins are"
      if b.advancers > b.decliners { return [.text("\(b.advancers) of \(total) \(coins) trading higher over the last 24h.")] }
      if b.decliners > b.advancers { return [.text("\(b.decliners) of \(total) \(coins) trading lower over the last 24h.")] }
      return [.text("Your coins are evenly split between gainers and losers over the last 24h.")]
    }
    return []
  }

  static func mergeAdjacentText(_ segments: [Segment]) -> [Segment] {
    var out: [Segment] = []
    for s in segments {
      if case .text(let t) = s, case .text(let prev)? = out.last { out[out.count - 1] = .text(prev + t) } else { out.append(s) }
    }
    return out
  }

  public static func buildSegments(portfolioChangePct: Double?, marketChangePct: Double?, sentiment: SentimentSummary?, breadth: BreadthStats?) -> [Segment] {
    let perf = performance(portfolio: portfolioChangePct, market: marketChangePct)
    let t = tone(sentiment: sentiment, breadth: breadth)
    var segs = perf
    if !perf.isEmpty && !t.isEmpty { segs.append(.text(" ")) }
    segs.append(contentsOf: t)
    return mergeAdjacentText(segs)
  }
}
